### ISSUE TITLE
The performance of DDP acceleration is lower than single GPU training.

### ISSUE THREAD

**Issue by JiuFengSC** [huggingface/accelerate#1277]
### System Info

```Shell
- `Accelerate` version: 0.15.0
- Platform: Linux-5.4.0-81-generic-x86_64-with-glibc2.17
- Python version: 3.8.13
- Numpy version: 1.23.5
- PyTorch version (GPU?): 1.13.1 (True)
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: fp16
        - use_cpu: False
        - dynamo_backend: NO
        - num_processes: 4
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: all
        - main_process_ip: None
        - main_process_port: None
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - deepspeed_config: {}
        - fsdp_config: {}
        - megatron_lm_config: {}
        - downcast_bf16: no
        - tpu_name: None
        - tpu_zone: None
        - command_file: None
        - commands: None
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I was finetuning the CLIP model with another dataset.
I used the HuggingFace Accelerate library to speed up, but the problem is that DDP training’s performance is lower than single GPU training. Some web pages said Batch Normalization might cause it, since every single batch is divided into different GPUs, we need to synchronize the BN value.
But there is no BN in CLIP/transformer-based architecture.

The Loss gap is like this figure Upper one is hugging face DDP and the bottom one is single GPU training. Also, there is a gap in performance.
<img width="416" alt="image" src="https://user-images.githubusercontent.com/80580308/229421395-c729cdec-0188-41f0-933a-a06cbf7753f0.png">

To simplify, here is only main() function and train() function. 

```python
def train(clip_model, train_dataloader, Max_Epoch, loss_img, loss_txt, optimizer, device, start_epoch=0):
    clip_model.train()
    logfile = open(args.exp_dir+"/finetuning_log.txt", "a+")
    time_str = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())
    logfile.write(' Start Training{}\n'.format(time_str))
    logfile.flush()
    for epoch in range(start_epoch, Max_Epoch):
        loss_epoch = 0
        counter = torch.tensor(0).to(device)
        # import ipdb;ipdb.set_trace()
        with tqdm(train_dataloader) as tepoch:
            for batch in tepoch:
                img, label_sentence = batch
                optimizer.zero_grad()
                torch.autograd.set_detect_anomaly(True)

                logits_per_image, logits_per_text = clip_model(img, label_sentence)

                ground_truth = torch.arange(len(img), dtype=torch.long).to(device)
                loss = (loss_img(logits_per_image, ground_truth)+loss_txt(logits_per_text, ground_truth))/2
                accelerator.backward(loss)

                convert_models_to_fp32(clip_model)
                optimizer.step() # Remove the mixed-precise from CLIP
                clip.model.convert_weights(clip_model)


                loss_epoch += loss.detach()
                counter += 1
                # bar_dict = {"loss":loss_epoch/counter,"epoch":epoch,"l_i":loss_of_img,"l_t":loss_of_txt}
                bar_dict = {"loss": loss_epoch.item()/counter.item(), "epoch": epoch}
                logfile.write(
                    f"Epoch[{epoch}] loss=={loss}\n")
                logfile.flush()
                tepoch.set_postfix(bar_dict)
                # print(type(total_loss))
        if epoch % 5 == 0:
            if not os.path.exists(f"{args.exp_dir}/checkpoints{epoch}"):
                os.makedirs(f"{args.exp_dir}/checkpoints{epoch}")
            accelerator.save_state(output_dir=f"{args.exp_dir}/checkpoints{epoch}")

    logfile.close()


if __name__ == '__main__':
    # Initialize CLIP
    # device = "cuda" if torch.cuda.is_available() else "cpu"
    accelerator = Accelerator()
    device = accelerator.device
    clip_model, preprocess = clip.load("ViT-B/32", device=device)

    
    if device == "cpu":
        clip_model.float()
    else:
        # clip.model.convert_weights(clip_model)
        clip_model.apply(_convert_weights_to_fp16)

    load_epoch = 0

    loss_img = nn.CrossEntropyLoss()
    loss_txt = nn.CrossEntropyLoss()
    optimizer = optim.Adam(clip_model.parameters(), lr=1e-7,
                        betas=(0.9, 0.98), eps=1e-8, weight_decay=0.001)


    audio_conf = {'num_mel_bins': 128, 'target_length': args.audio_length, 'freqm': args.freqm, 'timem': args.timem, 'mixup': args.mixup,
                'dataset': args.dataset, 'mode': 'train', 'mean': args.dataset_mean, 'std': args.dataset_std,'noise': args.noise,"skip_norm":args.skip_norm}

    dataset = VTDataset(args.data_train, preprocess, audio_conf, label_csv=args.label_csv,prompt=False)
    train_dataloader = DataLoader(dataset, batch_size=args.batch_size, num_workers=args.nDataLoaderThread, shuffle=True)

    clip_model, optimizer, train_dataloader = accelerator.prepare(clip_model, optimizer, train_dataloader)
    train(clip_model, train_dataloader, 100, loss_img, loss_txt, optimizer,device)

```
Even though I tried the `clip_model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(clip_model).to(device)
`, it still doesn't work.

### Expected behavior

The accelerated training should have a similar performance with single GPU.

**Comment by muellerzr** [comment]
Please read this doc, you need to be mindful of your batch size when doing such an analysis and learning rate. Let us know if after this there are still discrepancies https://huggingface.co/docs/accelerate/concept_guides/performance

**Comment by sentialx** [comment]
@muellerzr I followed the doc when pre-training OPT and here are my results:
![image](https://user-images.githubusercontent.com/11065386/229941799-226da825-5500-46c1-9caf-6c60ed9f6fbd.png)

Is that an expected behavior, or should the 2x run be identical to the 1x GPU?

**Comment by muellerzr** [comment]
@sentialx are you using a scheduler?

**Comment by sentialx** [comment]
Yeah I use transformers.get_constant_schedule_with_warmup

With 750 warmup steps, 20k total steps and base learning rate 6e-4 / 3

**Comment by muellerzr** [comment]
@sentialx how are you creating your Scheduler? Like are you taking into account the Gradient accumulation steps in the length? (You shouldn't)

**Comment by sentialx** [comment]
No, I don't, total_steps is a constant of 20k

```
    lr_scheduler = get_cosine_schedule_with_warmup(
        optimizer, num_warmup_steps=warmup, num_training_steps=total_steps)
```

**Comment by muellerzr** [comment]
Just ran some stuff myself so I could 100% get you on the right track @sentialx :) 

So first, *some* degradation in performance is expected with gradient accumulation, though it should be minimal.

Second, I wouldn't actually recommend dividing the loss, I just ran some tests to verify what would be 1:1 equivalent, and if your scheduler comes from `transformers` then dividing the expected batch size in half should be enough (vs on a full GPU).

Regarding current step, you can also just do:

```python
current_step = 0
if accelerator.sync_gradients:
    current_step += 1
     accelerator.print(torch.mean(accelerator.gather(avg_loss / total_accumulation_steps)))
```

Now, onto what is "wrong" with the code. Actually, nothing. This is expected. I also saw more differences when using lower precision. 

The reason for this is because I believe due to the averaging across GPUs on the gradients, the resulting gradient will never be exactly what it was on a GPU with less accumulation steps (or none). So, in regards to your code this is correct because you're using gradient accumulation.  You'll also notice the end result is very similar as well.

**Comment by muellerzr** [comment]
@sentialx @JiuFengSC I've opened a PR to adjust the docs based on my last reply and findings: https://github.com/huggingface/accelerate/pull/1296

**Comment by sentialx** [comment]
@muellerzr Thank you very much!
Just one more question:
Is the order of batches dispatched to multiple GPUs always guaranteed?
e.g.
my batch indices look like this (batch_size=3, accum_steps=2):
```
[0, 4, 8]
[12, 16, 20]
[1, 5, 9]
[13, 17, 21]
```

Is it guaranteed that cuda:0 will receive
```
[0, 4, 8]
[1, 5, 9]
```
and cuda:1 will receive
```
[12, 16, 20]
[13, 17, 21]
```
?
The order is important for my use case.
